### DISCUSSION TITLE
"I can has strict inequality?"

### DISCUSSION THREAD

**Discussion by zoharl3** [yalmip/YALMIP#1059] [Problems using YALMIP] [answered]
"No. Every time you use it a cat cries somewhere."

Matlab pops up a figure of a kitten with this (grammatically incorrect) message. Either it's a virus, or it's yalmip and maybe gurobi (not sure). Any idea?
I thought yalmip asserts on strict ineq?

**Comment by johanlofberg** [comment, marked as answer]
It is YALMIP telling you that you cannot have strict inequalities.

If you think it should be possible, please give us the solution to minimize x subject to x>0

**Reply by johanlofberg** [reply]
https://www.google.com/search?q=I+can+haz&sxsrf

**Reply by zoharl3** [reply]
I went over the whole code, and I finally found the place where I mistyped '<'. Call me old fashion, but I still prefer the old assert with a stack trace.

**Reply by johanlofberg** [reply]
The annoying cat was to prune out all the people who still had refused to fix code despite warnings for 10 years. It was added 1 year ago. In the develop version, i.e.,  in the next official release, it simply errors out completely. If you previously had an assert (i.e. error and code stopping), it means you have been running the develop branch.

**Reply by zoharl3** [reply]
Why is it an issue? I mean, just treat strict ineq as non-strict since the former is meaningless.

**Reply by johanlofberg** [reply]
There are a lot of people thinking you can have strict inequalities, and then when a non-strictly satisfied solution is returned, they are confused and/or report bugs. A soft change has been running for 15 years from silently just treating as non-strict, warning, annoying cat, and now (next version) simply not available as an operator.

**Reply by scientistl** [reply]
> It is YALMIP telling you that you cannot have strict inequalities.
> 
> If you think it should be possible, please give us the solution to minimize x subject to x>0

As suggested in the Yalmip docs https://yalmip.github.io/inside/strictinequalities/ (which you probably wrote), the way to solve a strict inequality is to solve a non strict version, and then verify it satisfies the strict condition. If the solution doesn't satisfy the strict condition, add a margin to the strict inequalities and solve with said margin. The magic margin should be a bit larger than the solver tolerance, though a factor of 10 should also be good.

There's series motivation for this. Whole scientific fields, like modern control theory, rely on strict solutions. To convexify problems we often define the "real variable" as va1/var2 with var2 > 0. If var2 is allowed to be even slightly infeasible, the solution becomes meaningless.

Of course, we can keep implementing these tolerance tests and adding margins on our own. It will improve usability if strict inequalities are allowed and implemented as you explained in the docs. Matlab makes it hard to add functions (requires a different file and many Matlab users would prefer emailing files to a Github repo). This means that not having to copy paste this solution would be very helpful.

@johanlofberg

**Reply by johanlofberg** [reply]
Adding a layer between a solver and the modelling language which performs various hacks to resolve problems searching for margins which would render a solution strictly feasible but not interfere with feasibility would open up for a world of problems. Much better that user understand the problem with defining strict inequalities and then use problem specific knowledge to deal with it. In the forthcoming release, strict inequalities simply aren't supported, i.e. the operators simply yields error.

**Reply by scientistl** [reply]
Errors would indeed be much better. A single missing = sign inside a loop can make my PC crash due to too many cat warnings (it happened). Please make the stable version return a (hopefully meaningful) error ASAP.
I can't see the world of problems you're referring to, so it may be worth adding an explanation about these problems in the docs, rather then ignoring this potential yet apparently problematic solution.
Also, thanks for the quick replay! @johanlofberg

**Reply by johanlofberg** [reply]
The problems would be models where there are many strict constraints, some of them infeasible, and then some hack built on top of that trying to figure if the problem actually is lacking an interior solution or if the solver simply is struggling. That mess would have to be done by the user, not a generic code in YALMIP which would be overly complicated. The annoying cat has been in the code to get people who have old legacy code to update their code before the code errors out, which they did not do when it only warned. We are talking code 20+ old here, so there are a lot of things to take into consideration. If you want to have the error already now, just run the develop branch.
